tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,244 @@
+use sharded_slab::{Guard, Slab};
+
+use crate::{
+    fmt::span::SpanStack,
+    registry::{
+        extensions::{Extensions, ExtensionsInner, ExtensionsMut},
+        LookupMetadata, LookupSpan, SpanData,
+    },
+    sync::RwLock,
+};
+use std::{
+    cell::RefCell,
+    sync::atomic::{fence, AtomicUsize, Ordering},
+};
+use tracing_core::{
+    dispatcher,
+    field::FieldSet,
+    span::{self, Current, Id},
+    Event, Interest, Metadata, Subscriber,
+};
+
+/// `Registry` is a shared, reusable store for spans.
+#[derive(Debug)]
+pub struct Registry {
+    spans: Slab<Data>,
+}
+
+/// The [Registry]'s view of a span. This definition
+/// is intentionally kept small—only data pertaining
+/// to span relationships, span metadata, and active
+/// references are stored. Additional data, such as
+/// formatted fields, are stored within the extensions
+/// typemap.
+#[derive(Debug)]
+pub struct Data {
+    metadata: &'static Metadata<'static>,
+    parent: Option<Id>,
+    ref_count: AtomicUsize,
+    pub(crate) extensions: RwLock<ExtensionsInner>,
+}
+
+// === impl Registry ===
+
+impl Default for Registry {
+    fn default() -> Self {
+        Self { spans: Slab::new() }
+    }
+}
+
+#[inline]
+fn idx_to_id(idx: usize) -> Id {
+    Id::from_u64(idx as u64 + 1)
+}
+
+#[inline]
+fn id_to_idx(id: &Id) -> usize {
+    id.into_u64() as usize - 1
+}
+
+impl Registry {
+    fn insert(&self, s: Data) -> Option<usize> {
+        self.spans.insert(s)
+    }
+
+    fn get(&self, id: &Id) -> Option<Guard<'_, Data>> {
+        self.spans.get(id_to_idx(id))
+    }
+}
+
+thread_local! {
+    static CURRENT_SPANS: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+impl Subscriber for Registry {
+    fn register_callsite(&self, _: &'static Metadata<'static>) -> Interest {
+        Interest::always()
+    }
+
+    fn enabled(&self, _: &Metadata<'_>) -> bool {
+        true
+    }
+
+    #[inline]
+    fn new_span(&self, attrs: &span::Attributes<'_>) -> span::Id {
+        let parent = if attrs.is_root() {
+            None
+        } else if attrs.is_contextual() {
+            self.current_span().id().map(|id| self.clone_span(id))
+        } else {
+            attrs.parent().map(|id| self.clone_span(id))
+        };
+
+        let s = Data {
+            metadata: attrs.metadata(),
+            parent,
+            ref_count: AtomicUsize::new(1),
+            extensions: RwLock::new(ExtensionsInner::new()),
+        };
+        let id = self.insert(s).expect("Unable to allocate another span");
+        idx_to_id(id)
+    }
+
+    /// This is intentionally not implemented, as recording fields
+    /// on a span is the responsibility of layers atop of this registry.
+    #[inline]
```

Comment:
I don't _think_ it's necessary to request this be inlined; since it's a totally empty function, I think it should be optimized away entirely in release mode.